### PR TITLE
Issue 28 fao manage custom vocabularies

### PR DESCRIPTION
**Pull request by abulojoshua1**
### What does the PR do?
- Adds functionality to manage custom vocabularies from CKAN UI
- Localizes custom vocabulary tags (see [issue #24](https://github.com/geosolutions-it/ckanext-faoclh/issues/24))

### Changes made
- Added a tab to manage CKAN vocabulary admin UI

### Issue
[issue #28](https://github.com/geosolutions-it/ckanext-faoclh/issues/28)
[issue #24](https://github.com/geosolutions-it/ckanext-faoclh/issues/24)

### related pull request
- https://github.com/geosolutions-it/ckanext-multilang/pull/22

### CODE REVIEW

**Comment by etj** [comment]
Shouldnt the button at th ebottom be "Save tag" or "update tag"?

image

**Comment by etj** [comment]
1) Full list
image
2) select one item:
image
3) fill for instance english field, and press the "create voc":
image
Problems:
- The edited text is lost
- The error says the original name was worng in first place!

File: ckanext/faoclh/controller.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            u'fao_resource_type': u'Resource Type',
+            u'fao_activity_type': u'Activity Type',
+            u'fao_geographic_focus': u'Geographical Focus'
+        }
+        self.errors = []
+        self.available_locales = get_locales_from_config()
+        user = toolkit.get_action('get_site_user')({'ignore_auth': True}, {})
+        self.context = {'user': user['name'], 'ignore_auth': True}
+
+    def vocabularies(self, *args, **kwargs):
```
Comment (etj) [inline ckanext/faoclh/controller.py:32]: This method is used for both GET and POST, can't we have two different methods instead? Code would be much cleaner

File: ckanext/faoclh/controller.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            result = toolkit.get_action(u'tag_create')(self.context, data)
+            self.localize_tags(result.get('id'))
+            self.created = True
+            return result.get('id')
+        except ValidationError as e:
+            if u'already belongs to vocabulary' in str(e):
+                self.errors.append(_(u'Vocabulary tag is already existent.'))
+                self.created = False
+            return e
+        return result
```
Comment (etj) [inline ckanext/faoclh/controller.py:93]: Instead of using an instance variable for `created`, consider using
```
   return result, created
```

**Comment by etj** [comment]
What's the "Cancel" button about?
image

Every time I press it, I get this error:
image

Is it a "delete" action?

File: ckanext/faoclh/controller.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+log = logging.getLogger(__name__)
+
+
+class VocabularyController(AdminController):
+    def __init__(self):
+        self.created = False
+        self.template_mapper = {
+            u'fao_resource_type': u'Resource Type',
+            u'fao_activity_type': u'Activity Type',
+            u'fao_geographic_focus': u'Geographical Focus'
```
[the remark this answers is not in the document]
Comment (abulojoshua1) [reply ckanext/faoclh/controller.py:25]: Using centralized constants will mean that I'll have to write conditional statements to change `vocab_name_name` in templates depending on `vocab_name` from the HTTP request which will be a bit unclean.

File: ckanext/faoclh/controller.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        self.context = {'user': user['name'], 'ignore_auth': True}
+
+    def vocabularies(self, *args, **kwargs):
+        self.request = kwargs[u'pylons'].request
+        self.name = self.request.POST.get(u'name', u'')
+        self.lang = helpers.getLanguage()
+        self.vocab_name = self.request.GET.get(
+            u'vocab_name', u'fao_resource_type')
+        self.tag = self.request.GET.get(u'tag', u'')
+        self.vocabs = self.get_vocab({}, self.vocab_name)
```
[the remark this answers is not in the document]
Comment (abulojoshua1) [reply ckanext/faoclh/controller.py:39]: Your missing nothing, I'll reduce instance fields and add them to the constructor just for more readability.

File: ckanext/faoclh/controller.py (hunk 1 above)
Comment (abulojoshua1) [reply ckanext/faoclh/controller.py:32]: Ok, I'll split the method for POST and GET.

**Comment by abulojoshua1** [comment]
> 1. Full list
>    image
> 2. select one item:
>    image
> 3. fill for instance english field, and press the "create voc":
>    image
>    Problems:
> 
> * The edited text is lost
> * The error says the original name was worng in first place!

@etj, the error you are getting is one being generated by vocabulary validations that I came across within CKAN while reading through the code. It does not allow for Upper-case characters, should I instead write code that validates the user entries instead or is there a vocabulary validation class/method that I may have missed?

**Comment by etj** [comment]
> > 1. Full list
> >    image
> > 2. select one item:
> >    image
> > 3. fill for instance english field, and press the "create voc":
> >    image
> >    Problems:
> > 
> > 
> > * The edited text is lost
> > * The error says the original name was worng in first place!
> 
> @etj, the error you are getting is one being generated by vocabulary validations that I came across within CKAN while reading through the code. It does not allow for Upper-case characters, should I instead write code that validates the user entries instead or is there a vocabulary validation class/method that I may have missed?

Are the json files used for initialization stil lused? We may fix those. 
Also, those files contain also English translations. We should be able to ingest them,

**Comment by etj** [comment]
@tdipisa  is it ok to force the editing of all the languages at once?
Different languages may be edited by different admins?

image

**Comment by etj** [comment]
Sometimes tags are not localized in the dataset list, while they are in the editing form.

image

**Comment by etj** [comment]
@tdipisa is it ok we only show items that have a localization?
It means: should we assume all the tags should be localized in all languages, or should we provide a default value (e.g. the key value) for the missing translations?

image

File: ckanext/faoclh/plugin.py [hunk 5]
```diff
@@ -209,31 +242,33 @@ def get_helpers(self):
...
 
 def fao_voc(voc_name):
-    path = config.get('fao.vocab.path')
-    vocab_file = os.path.join(path, voc_name + ".json")
+    tag_dict = {}
+    try:
+        data = {u'id': voc_name}
+        tags = toolkit.get_action(u'vocabulary_show')({}, data).get('tags', [])
+        for tag in tags:
+            tag_dict[tag.get('id')] = tag.get('display_name')
```
Comment (etj) [inline ckanext/faoclh/plugin.py:250]: maybe a dict comprehension would be better here

**Comment by etj** [comment]
If the "Save Vocabulary" (shouldnt it be "save tag" or "Save Vocabulary Item"? )  is pressed more than once, both when creating and updating the item, multiple tags will be saved:

image

image

**Comment by tdipisa** [comment]
> @tdipisa is it ok we only show items that have a localization?
> It means: should we assume all the tags should be localized in all languages, or should we provide a default value (e.g. the key value) for the missing translations?
> 
> image

@etj @abulojoshua1 I would like to have a default value for not localized items (e.g. the key value) so that the admin is able to use them too and properly identify if some translations is missing and so add it.

**Comment by tdipisa** [comment]
> @tdipisa is it ok to force the editing of all the languages at once?
> Different languages may be edited by different admins?
> 
> image

@etj @abulojoshua1, I'm ok with this: **force the editing of all the languages at once?**
that for me means on making all translation fields mandatory.

**Pull request merged**